### ISSUE TITLE
Install process is completely broken in presence of existing tensorflow install, and breaks that also

### ISSUE THREAD

**Issue by amit-jain-SBG** [google/nucleus#25]
-start with a blank machine
-install tensorflow 2

$`pip3 install --user google-nucleus==0.5.0`
```
Building wheels for collected packages: google-nucleus
  Building wheel for google-nucleus (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-gd0b47d0/google-nucleus/setup.py'"'"'; __file__='"'"'/tmp/pip-install-gd0b47d0/google-nucleus/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-fjnjhmg0 --python-tag cp38
       cwd: /tmp/pip-install-gd0b47d0/google-nucleus/
  Complete output (3 lines):
  /usr/lib/python3.8/site-packages/setuptools/version.py:1: UserWarning: Module google was already imported from /home/amit/.local/lib/python3.8/site-packages/google/__init__.py, but /tmp/pip-install-gd0b47d0/google-nucleus is being added to sys.path
    import pkg_resources
  This package does not support wheel creation.
  ----------------------------------------
  ERROR: Failed building wheel for google-nucleus
  Running setup.py clean for google-nucleus
Failed to build google-nucleus
Installing collected packages: google-nucleus
    Running setup.py install for google-nucleus ... done
  WARNING: Could not find .egg-info directory in install record for google-nucleus==0.5.0 from https://files.pythonhosted.org/packages/28/04/5da4ba708671d62100d7df3ce053718af58c4635080ef10d35dac7cc7c5e/google_nucleus-0.5.0.tar.gz#sha256=74b1a280a67f03c2ff751c5dcbc6a902ca5be590360b02086e8f8b55d963acfd
Successfully installed google-nucleus
```
There is a closed log explaining that this is OK.

However, 
`import tensorflow as tf`
```
File "~/.local/lib/python3.8/site-packages/google/protobuf/descriptor.py", line 530, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors should not be created directly, but only retrieved from their parent.
```
`from nucleus.io import vcf`
```
File "~/.local/lib/python3.8/site-packages/google/protobuf/descriptor.py", line 530, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors should not be created directly, but only retrieved from their parent.
```
Further,
$`pip3 uninstall google-nucleus`
```
WARNING: Skipping google-nucleus as it is not installed.
```

**Comment by pichuan** [comment]
Hi,
I wasn't quite able to reproduce your error. Maybe I have a different environment. But I can show you what I did:

I'm testing on GCP. First, I got a machine for testing:
```
gcloud compute instances create "${USER}-nucleus-16" \
  --scopes "compute-rw,storage-full,cloud-platform"  \
  --image-family="ubuntu-1604-lts" \
  --image-project="ubuntu-os-cloud" \
  --machine-type "n1-standard-1"   --zone "us-west1-b"
```

ssh into the machine, and then, check Python3 version:

```
$ python3 --version
Python 3.5.2
```

I had to install pip on my machine:
```
curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
sudo python3 get-pip.py
python3 -m pip install --upgrade pip
```

Check pip3 version:
```
pip3 --version
pip 19.3.1 from /usr/local/lib/python3.5/dist-packages/pip (python 3.5)
```

Install TF2 and Nucleus:
```
pip3 install --user --upgrade  tensorflow==2.0.0
pip3 install --user google-nucleus==0.5.0
```

Then, I ran these two commands, and both of them finished without any errors:
```
python3 -c 'import tensorflow as tf'
python3 -c 'from nucleus.io import vcf'
```

@amit-jain-SBG if you can give me more information on your environment (OS version, python and pip version, etc), I can try again and see if I can reproduce your issue. 


There is one issue that I can confirm -- `pip3 uninstall` didn't work for me either:

```
$ pip3 uninstall google-nucleus
WARNING: Skipping google-nucleus as it is not installed.
```

I tried to check the list using `pip3 list`, and google-nucleus wasn't here. I'll check if this is a known issue. If not, I'll file an internal issue to track it.

**Comment by pichuan** [comment]
Hi @amit-jain-SBG ,
I checked with @ThomasColthurst , and here is what I learned:

The fact that google-nucleus doesn't exist in `pip list` (and can't be uninstalled) is a known issue.
The basic issue here is that while pip install supports installing from both wheel files and tar files, only wheel files are "first class citizens" -- tar files don't support pip list or pip uninstall.  But we have to use tar files because we need symbolic links and wheel files don't support those.

If you need to uninstall nucleus, you'll need to find where the package is installed and manually remove them.
Thanks.

I'll close this issue for now because I can't reproduce the error. If you can provide more information, please feel free to re-open and bug.

**Comment by amit-jain-SBG** [comment]
Hello @pichuan,

Thanks for investigating this. I am aware of issue #18 that describes the uninstall and install problem with pip.

However, incompatibility with TensorFlow is critical, in my opinion. I suggest you try with the latest Python (3.8+) - I have see reports of other packages experiencing install problems because of underlying changes introduced by this version. For example, https://github.com/pysam-developers/pysam/issues/860

For your information, I supply the relevant versions of things on my machine:
**System**: `Linux Burt 5.4.1-arch1-1-macbook #1 SMP PREEMPT x86_64 GNU/Linux`
**Python**: `Python 3.8.1 (default, Dec 21 2019, 20:57:38) [GCC 9.2.0] on linux`
**Tensorflow**: `python-tensorflow 2.1.0rc2-1`
**Protobuf**: `3.11.1-2`

**Comment by pichuan** [comment]
Hi @amit-jain-SBG 
After installing Python 3.8, I wasn't able to pip install tensorflow anymore.
I found this issue (still ongoing) on tensorflow that shows installing with python 3.8 is not supported yet: https://github.com/tensorflow/tensorflow/issues/33374

So I tested without installing tensorflow at all to see whether I still have issues with Nucleus itself (under Python 3.8). It seems like I am able to reproduce the issue:

```
pichuan@pichuan-nucleus-16:~$ python3.8 -c 'from nucleus.io import vcf'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/pichuan/.local/lib/python3.8/site-packages/nucleus/io/vcf.py", line 68, in <module>
    from nucleus.protos import variants_pb2
  File "/home/pichuan/.local/lib/python3.8/site-packages/nucleus/protos/variants_pb2.py", line 16, in <module>
    from nucleus.protos import reference_pb2 as nucleus_dot_protos_dot_reference__pb2
  File "/home/pichuan/.local/lib/python3.8/site-packages/nucleus/protos/reference_pb2.py", line 16, in <module>
    from nucleus.protos import range_pb2 as nucleus_dot_protos_dot_range__pb2
  File "/home/pichuan/.local/lib/python3.8/site-packages/nucleus/protos/range_pb2.py", line 36, in <module>
    _descriptor.FieldDescriptor(
  File "/home/pichuan/.local/lib/python3.8/site-packages/google/protobuf/descriptor.py", line 530, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors should not be created directly, but only retrieved from their parent.
```

Versions:
```
$ python3.8 --version
Python 3.8.1
$ pip3 --version
pip 19.3.1 from /usr/local/lib/python3.8/dist-packages/pip (python 3.8)
```

I have confirmed that Python 3.7 works, though.

I think this will take some time for us to resolve on our side. One short-term fix I can do is to clarify in the documentation that we support Python3 up to 3.7.
I'll add an internal issue to report that Nucleus doesn't yet work with Python 3.8.

**Comment by amit-jain-SBG** [comment]
OK. I believe I installed TF 2.1RC2-1 as a pre-built package available in Arch community repository.
Works fine on 3.8.

I agree with the fix you propose. 

But another piece of info that maybe useful for your team:

on 3.8 with TF installed and working, when I pip installed nucleus, TF stopped working (error described in the issue description). When I removed nucleus (i.e. delete sitepackages as suggested in #18 , TF started working again.

I leave it to you guys to figure out the compatibility of an install strategy that involves hacking the user's protobuf libraries with the aim of making a stable and reliable mechanism for genomic data transfer :)
